src/payload.iced

```diff
@@ -0,0 +1,37 @@
+{prng} = require('crypto')
+{createHash} = require('crypto')
+{createHmac} = require('crypto')
+msgpack = require('purepack')
+crypto = require('keybase-nacl')
+nonce = require('./nonce.iced')
+util = require('./util.iced')
+
+compute_authenticator = (hash, key) ->
+  hmac = createHmac('sha512', key)
+  hmac.update(hash)
+  return hmac.digest()[0...32]
+
+step1 = (header_hash, block_num, payload_secretbox) ->
+  step1_cat = Buffer.concat([header_hash, nonce.nonceForChunkSecretBox(block_num), payload_secretbox])
+  crypto_hash = createHash('sha512')
+  crypto_hash.update(step1_cat)
+  step1_hash = crypto_hash.digest()
+
+exports.generate_encryption_payload_packet = (payload_secretbox, block_num, header_hash, mac_keys) ->
+  step1_hash = step1(header_hash, block_num, payload_secretbox)
+
+  authenticators = []
+  for mac_key in mac_keys
+    authenticator = compute_authenticator(step1_hash, mac_key)
+    authenticators.push(authenticator)
+
+  payload_list = [authenticators, payload_secretbox]
+  return payload_list
+
+exports.parse_encryption_payload_packet = (payload_list, block_num, header_hash, mac_key, auth_index) ->
```

Comment:
This function checks the authenticator but doesn't actually decrypt the data. I'd prefer to do both in the same step, if we can.

The current approach means that the caller is going to be decrypting the data, but not checking the authenticator. Which is to say, the caller believes "the authenticator for this has already been checked, so I don't need to check it." I'd prefer to avoid ever writing code that has to make assumptions like that. Stuff might get refactored around in the future, and that assumption might not be true anymore, and if we forget to update all existing callers then we'd have a security issue.

When we were putting together the spec, I was hoping we could arrange things so that it was actually _impossible_ to forget the authentication step. (Like, maybe authentication was implicit in how you computed the nonce or something like that.) We didn't end up finding a way, but the next best thing is to make sure we always do authentication right next to decryption in all of our implementations.